### PR TITLE
feat(tdigest): add the support of TDIGEST.BYRANK and TDIGEST.BYREVRANK command

### LINKED ISSUE #3064: TDigest: Implement `TDIGEST.BYRANK` command
### Search before asking

- [x] I had searched in the [issues](https://github.com/apache/kvrocks/issues) and found no similar issues.

### Motivation

Refer: https://github.com/apache/kvrocks/issues/2558
We should implement `TDIGEST.BYRANK` command.

Redis command reference:
https://redis.io/docs/latest/commands/tdigest.byrank/

### Solution

Depends on #2791 #2792.

Follow the implementation of #2794 for this command.

### Are you willing to submit a PR?

- [x] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by donghao526**
This PR close #3064 and #3065 
1. Add TDIGEST.BYRANK and TDIGEST.BYREVRANK command implementation
2. Add Related cpp and go unit tests

### CODE REVIEW

**Comment by LindaSummer** [comment]
Hi @donghao526 ,

Thanks very much for your great effort! ❤️

I will review this PR in two days. 😊

Best Regards,
Edward

**Review by khanbilal732 — APPROVED** [review]

**Review by khanbilal732 — APPROVED** [review]

File: src/commands/cmd_tdigest.cc [hunk 1]
```diff
@@ -238,6 +238,76 @@ class CommandTDigestRevRank : public TDigestRankCommand<true> {};
...
+    std::vector<std::string> ranks;
+    ranks.reserve(origin_inputs_.size());
+    auto is_resp3 = conn->GetProtocolVersion() == RESP::v3;
+    for (const auto &v : origin_inputs_) {
+      if (is_resp3) {
+        ranks.push_back(redis::Double(redis::RESP::v3, result[unique_inputs_order_[v]]));
+      } else {
+        ranks.push_back(redis::BulkString(fmt::format("{}", result[unique_inputs_order_[v]])));
+      }
+    }
```
Comment (LindaSummer) [inline src/commands/cmd_tdigest.cc:296]:
```suggestion
    for (const auto &v : origin_inputs_) {
      auto rank_value = result[unique_inputs_order_[v]];
      if (is_resp3) {
        ranks.push_back(redis::Double(redis::RESP::v3, rank_value));
      } else {
        ranks.push_back(redis::BulkString(fmt::format("{}", rank_value)));
      }
    }
```

Maybe this way could be cleaner.

File: src/types/redis_tdigest.h [hunk 2]
```diff
@@ -81,6 +81,10 @@ class TDigest : public SubKeyScanner {
                        std::vector<int>& result);
   rocksdb::Status RevRank(engine::Context& ctx, const Slice& digest_name, const std::vector<double>& inputs,
                           std::vector<int>& result);
+  rocksdb::Status ByRevRank(engine::Context& ctx, const Slice& digest_name, const std::vector<int>& inputs,
```
Comment (LindaSummer) [inline src/types/redis_tdigest.h:84]: This is an optional advice, we'd better keep same style in this file and make an output value in parameter a pointer.

File: tests/cppunit/types/tdigest_test.cc [hunk 3]
```diff
@@ -448,4 +448,79 @@ TEST_F(RedisTDigestTest, RevRank_and_Rank_on_different_or_same_and_unordered_inp
...
+  }
+
+  // Test 3: Test boundary conditions
+  std::vector<int> boundary_ranks = {0, 7, 14, 100};
+  result.clear();
+  status = tdigest_->ByRank(*ctx_, test_digest_name, boundary_ranks, result);
+  ASSERT_TRUE(status.ok()) << status.ToString();
+  EXPECT_EQ(result[0], 1.0) << "Rank 0 should be minimum";
+  EXPECT_TRUE(std::isinf(result[3])) << "Rank >= total_weight should be infinity";
 }
```
Comment (LindaSummer) [inline tests/cppunit/types/tdigest_test.cc]:
```suggestion
}

```

File: tests/gocase/unit/type/tdigest/tdigest_test.go [hunk 4]
```diff
@@ -75,6 +101,11 @@ func TestTDigest(t *testing.T) {
 			Options:    []string{"yes", "no"},
 			ConfigType: util.YesNo,
 		},
+		{
```
Comment (LindaSummer) [inline tests/gocase/unit/type/tdigest/tdigest_test.go]: Do we need this for all tdigest test cases or just for this command?
If we need to enable for all commands, maybe a new PR will be better.

File: tests/gocase/unit/type/tdigest/tdigest_test.go [hunk 5]
```diff
@@ -716,4 +747,231 @@ func tdigestTests(t *testing.T, configs util.KvrocksServerConfigs) {
...
+		require.NoError(t, rsp.Err())
+		vals, err := rsp.Slice()
+		require.NoError(t, err)
+		require.Len(t, vals, 8)
+		// test BYRANK with duplicate values
+		if isRESP3 {
+			for i, v := range vals {
+				rank, ok := v.(float64)
+				require.True(t, ok, "expected float64 but got %T at index %d", v, i)
+				expected := []float64{1, 2, 2, 3, 3, 3}
```
Comment (LindaSummer) [inline tests/gocase/unit/type/tdigest/tdigest_test.go]: It should be out of the for loop.

File: tests/gocase/unit/type/tdigest/tdigest_test.go [hunk 6]
```diff
@@ -716,4 +747,231 @@ func tdigestTests(t *testing.T, configs util.KvrocksServerConfigs) {
...
+		rsp = rdb.Do(ctx, "TDIGEST.BYREVRANK", key, "0", "1", "2", "3", "4", "5", "6", "7")
+		require.NoError(t, rsp.Err())
+		vals, err = rsp.Slice()
+		require.NoError(t, err)
+		require.Len(t, vals, 8)
+		if isRESP3 {
+			for i, v := range vals {
+				rank, ok := v.(float64)
+				require.True(t, ok, "expected float64 but got %T at index %d", v, i)
+				expected := []float64{3, 3, 3, 2, 2, 1}
```
Comment (LindaSummer) [inline tests/gocase/unit/type/tdigest/tdigest_test.go]: Same as above.

File: tests/gocase/unit/type/tdigest/tdigest_test.go [hunk 7]
```diff
@@ -716,4 +747,231 @@ func tdigestTests(t *testing.T, configs util.KvrocksServerConfigs) {
...
+		rsp := rdb.Do(ctx, "TDIGEST.BYRANK", key, "0", "1", "2", "3", "4", "5", "6", "7", "7", "5", "20", "100")
+		require.NoError(t, rsp.Err())
+		vals, err := rsp.Slice()
+		require.NoError(t, err)
+		require.Len(t, vals, 12)
+		if isRESP3 {
+			for i, v := range vals {
+				rank, ok := v.(float64)
+				require.True(t, ok, "expected float64 but got %T at index %d", v, i)
+				expected := []float64{1, 4, 7, 8.8, 8.8, 12, 12, 15, 15, 12}
```
Comment (LindaSummer) [inline tests/gocase/unit/type/tdigest/tdigest_test.go]: Same as above.

File: tests/gocase/unit/type/tdigest/tdigest_test.go [hunk 8]
```diff
@@ -716,4 +747,231 @@ func tdigestTests(t *testing.T, configs util.KvrocksServerConfigs) {
...
+		rsp = rdb.Do(ctx, "TDIGEST.BYREVRANK", key, "20", "75", "0", "1", "2", "3", "4", "5", "6", "7")
+		require.NoError(t, rsp.Err())
+		vals, err = rsp.Slice()
+		require.NoError(t, err)
+		require.Len(t, vals, 10)
+		if isRESP3 {
+			for i, v := range vals {
+				rank, ok := v.(float64)
+				require.True(t, ok, "expected float64 but got %T at index %d", v, i)
+				expected := []float64{0, 0, 100, 99, 81, 75, 50, 46, 36, 36}
```
Comment (LindaSummer) [inline tests/gocase/unit/type/tdigest/tdigest_test.go]: Same as above.

Comment (LindaSummer) [reply tests/gocase/unit/type/tdigest/tdigest_test.go]: In redis lastes docker image I find that the result is below. The last one seems a little different.
```
TDIGEST.BYREVRANK test 20 75 0 1 2 3 4 5 6 7
 1) "-inf"
 2) "-inf"
 3) "100"
 4) "99"
 5) "81"
 6) "75"
 7) "50"
 8) "46"
 9) "36"
10) "32.5"
```

**Review by LindaSummer — COMMENTED** [review]
Hi @donghao526 ,

Left some comments.
Generally LGTM. 😊
Thanks for your effort!

File: tests/gocase/unit/type/tdigest/tdigest_test.go (hunk 4 above)
Comment (donghao526) [reply tests/gocase/unit/type/tdigest/tdigest_test.go]: Just for this command, I will make this change command-specific.

File: tests/gocase/unit/type/tdigest/tdigest_test.go (hunk 8 above)
Comment (donghao526) [reply tests/gocase/unit/type/tdigest/tdigest_test.go]: Is your test sketch the same as this test?
I test on cloud.redis.io (Redis 8.2) and got 36. I'll double-check with a Docker image later
Image

Comment (donghao526) [reply tests/gocase/unit/type/tdigest/tdigest_test.go]:
```
[donghao526@VM-165-13~]$ sudo docker run -d --name redis -p 6379:6379 redis:latest
2d77559739849b9fee49a86333ca5226fc5afdb8f46bf06bc318d09ca006f3b4   
[donghao526@VM-165-13~]$ redis-cli -h 127.0.0.1 -p 6379
127.0.0.1:6379> TDIGEST.CREATE t COMPRESSION 1000
OK
127.0.0.1:6379> TDIGEST.ADD t 12 100 50 36 75 81 35.5 46 36 8.8 15 4 32.5 12 8.8 7 99 1 
OK
127.0.0.1:6379> tdigest.byrevrank t 20 75 0 1 2 3 4 5 6 7
 1) "-inf"
 2) "-inf"
 3) "100"
 4) "99"
 5) "81"
 6) "75"
 7) "50"
 8) "46"
 9) "36"
10) "36"
```
Test with the latest docker image, the last one is "36"

**Review by khanbilal732 — APPROVED** [review]

**Comment by donghao526** [comment]
> Hi @donghao526 ,
> 
> Left some comments. Generally LGTM. 😊 Thanks for your effort!

Hi @LindaSummer 
Thanks for you review. I have resolved these comments. ❤️

**Comment by LindaSummer** [comment]
> > Hi @donghao526 ,
> 
> > 
> 
> > Left some comments. Generally LGTM. 😊 Thanks for your effort!
> 
> 
> 
> Hi @LindaSummer 
> 
> Thanks for you review. I have resolved these comments. ❤️

got it!🚀
I will go through this PR later today.

**Review by khanbilal732 — APPROVED** [review]

File: tests/gocase/unit/type/tdigest/tdigest_test.go (hunk 8 above)
Comment (LindaSummer) [reply tests/gocase/unit/type/tdigest/tdigest_test.go]: Hi @donghao526 ,

I have double-checked and find that I input with a wrong number.
Sorry for my mistake in review.

File: tests/gocase/unit/type/tdigest/tdigest_test.go [hunk 9]
```diff
@@ -54,17 +55,42 @@ type tdigestInfo struct {
...
-		Capacity:          v["Capacity"].(int64),
-		MergedNodes:       v["Merged nodes"].(int64),
-		UnmergedNodes:     v["Unmerged nodes"].(int64),
-		MergedWeight:      v["Merged weight"].(int64),
-		UnmergedWeight:    v["Unmerged weight"].(int64),
-		Observations:      v["Observations"].(int64),
-		TotalCompressions: v["Total compressions"].(int64),
+	// Handle both RESP2 (map) and RESP3 (array) formats
+	switch v := value.(type) {
+	case map[interface{}]interface{}:
```
Comment (LindaSummer) [inline tests/gocase/unit/type/tdigest/tdigest_test.go]: This seems a little tricky, we'd better pass a parameter explicitly. Or we could provide two versions and let the test case choose which version should use.

**Review by LindaSummer — COMMENTED** [review]
Left one comment.
Generally LGTM.
Thanks very much for your effort! 😊

File: tests/gocase/unit/type/tdigest/tdigest_test.go (hunk 9 above)
Comment (donghao526) [reply tests/gocase/unit/type/tdigest/tdigest_test.go]: Got it. I've reverted this part since the RESP3 test only applies to byrank commands. It was an oversight that I didn't remove it before.

**Review by LindaSummer — APPROVED** [review]
Hi @donghao526 ,
LGTM.
Thanks very much for your contribution!❤️
